### PR TITLE
DSWS_S1: Add output validation #285

### PR DESCRIPTION
**Pull request by JimHofman**
## Description
- Add output verification to dswx-s1

## Testing
- Added unit tests 

## Comment
- Removed all reference to product_id

### CODE REVIEW

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 1]
```diff
@@ -112,28 +110,87 @@ class DSWxS1PostProcessorMixin(PostProcessorMixin):
...
+            self.logger.critical(self.name, ErrorCode.OUTPUT_NOT_FOUND, error_msg)
+
+        for out_product in output_products:
+            if not getsize(out_product):
+                error_msg = f"SAS output file {out_product} was created, but is empty"
+
+                self.logger.critical(self.name, ErrorCode.INVALID_OUTPUT, error_msg)
+
+            #  Gather the burst output files into a dictionary
+            #     key = burst type (e.g. B01_WTR.tif)
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:155]: Bit of a misnomer here, should say "band type", not "burst type"

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 2]
```diff
@@ -112,28 +110,87 @@ class DSWxS1PostProcessorMixin(PostProcessorMixin):
...
+            #     value = list of filenames of this type (e.g. ['OPERA_L3_DSWx-S1_..._v0.1_B01_WTR.tif', ...]
+            if output_extension in out_product:
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:157]: Is this if statement necessary? You've already filtered the list of `out_products` to those that contain `output_extension` in the file name?

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 3]
```diff
@@ -112,28 +110,87 @@ class DSWxS1PostProcessorMixin(PostProcessorMixin):
...
+        """
+        Evaluates the output file(s) generated from SAS execution to ensure:
+            - That the file(s) contains some content (size is greater than 0).
+            - That the .tif output files (burst data) end with 'B01_WTR',
+              'B02_BWTR', or 'B03_CONF'
+            - That the there are the same number of each type of file, implying
+              3 output bands per tile
+
+        """
+        burst_dict = {}
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:130]: This should probably be named "band_dict" based on how it is used.

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 4]
```diff
@@ -112,28 +110,87 @@ class DSWxS1PostProcessorMixin(PostProcessorMixin):
...
+            #  Gather the burst output files into a dictionary
+            #     key = burst type (e.g. B01_WTR.tif)
+            #     value = list of filenames of this type (e.g. ['OPERA_L3_DSWx-S1_..._v0.1_B01_WTR.tif', ...]
+            if output_extension in out_product:
+                key = '_'.join(out_product.split('_')[-2:])
+                if key not in burst_dict:
+                    burst_dict[key] = []
+                burst_dict[key].append(out_product)
+
+        if len(burst_dict.keys()) != 3:
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:163]: Let's assign `3` to a constant, such as `EXPECTED_NUM_BANDS`

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 5]
```diff
@@ -112,28 +110,87 @@ class DSWxS1PostProcessorMixin(PostProcessorMixin):
...
+        Evaluates the output file(s) generated from SAS execution to ensure:
+            - That the file(s) contains some content (size is greater than 0).
+            - That the .tif output files (burst data) end with 'B01_WTR',
+              'B02_BWTR', or 'B03_CONF'
+            - That the there are the same number of each type of file, implying
+              3 output bands per tile
+
+        """
+        burst_dict = {}
+        num_bursts = []
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:131]: Similarly, this should be `num_bands`

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 6]
```diff
@@ -112,28 +110,87 @@ class DSWxS1PostProcessorMixin(PostProcessorMixin):
...
+
+        if len(burst_dict.keys()) != 3:
+            error_msg = f"Invalid SAS output file, too many burst types: {burst_dict.keys()}"
+
+            self.logger.critical(self.name, ErrorCode.INVALID_OUTPUT, error_msg)
+
+        # Make a list of the numbers of bursts per burst type
+        for burst in burst_dict.keys():
+            num_bursts.append(len(burst_dict[burst]))
+        if not all(burst_type == num_bursts[0] for burst_type in num_bursts):
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:171]: I'm a little confused by this if statement, `num_bursts` is a list so it does not make sense to iterate over `burst_type`. Did you mean to use `burst_dict`? Also, `num_bursts[0]` would be the number of files for a type, so why is it compared to `burst_type`?

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 7]
```diff
@@ -147,6 +204,7 @@ class DSWxS1Executor(DSWxS1PreProcessorMixin, DSWxS1PostProcessorMixin, PgeExecu
 
     NAME = "DSWx"
     """Short name for the DSWx-S1 PGE"""
+    # TODO do we not want to differentiate between S1 and HLS
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:207]: Not sure what this TODO is in reference to, can it be removed? Also, `NAME` should be set to `DSWx-S1`.

File: src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py [hunk 8]
```diff
@@ -141,6 +155,41 @@ def _compare_algorithm_parameters_runconfig_to_expected(self, runconfig):
...
+        # Clears old burst files (default 'True')
+        output_dir = join(self.test_dir, str(self.test_output_dir))
+        if clear:
+            path = output_dir
+            cmd = f"rm {path}/*.tif"
+            os.system(cmd)
+        # Add files to the output directory
+        for burst_output_file in burst_data:
+            if not empty_file:
+                # os.system(f"echo 'Test data string' >> {join(self.test_output_dir, burst_output_file)}")
```
Comment (collinss-jpl) [inline src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py:187]: Lets remove these commented out lines if they're no longer necessary

File: src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py [hunk 9]
```diff
@@ -141,6 +155,41 @@ def _compare_algorithm_parameters_runconfig_to_expected(self, runconfig):
...
+            Clear the output directory before writing new files (default=True)
+
+        """
+        # example of burst data passed to method:
+        # burst_data = ('OPERA_L3_DSWx-S1_burst_1_B01_WTR.tif', 'OPERA_L3_DSWx-S1_burst_1_B02_BWTR.tif',
+        #               'OPERA_L3_DSWx-S1_burst_1_B03_CONF.tif', 'OPERA_L3_DSWx-S1_burst_2_B01_WTR.tif',
+        #               'OPERA_L3_DSWx-S1_burst_2_B02_BWTR.tif', 'OPERA_L3_DSWx-S1_burst_2_B03_CONF.tif')
+
+        # Clears old burst files (default 'True')
+        output_dir = join(self.test_dir, str(self.test_output_dir))
```
Comment (collinss-jpl) [inline src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py:179]: I don't think you need this `join` here, `self.test_output_dir` should already be set to the path you're looking for.

File: src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py [hunk 10]
```diff
@@ -518,6 +568,127 @@ def test_dswx_s1_pge_input_validation(self):
...
+            yaml.safe_dump(runconfig_dict, config_fh, sort_keys=False)
+
+        try:
+            pge = DSWxS1Executor(pge_name="DSWxS1PgeTest", runconfig_path=test_runconfig_path)
+
+            with self.assertRaises(RuntimeError):
+                pge.run()
+
+            expected_output_file = 'dswx_s1_pge_test/output_dir/missing_dswx_s1.tif'
+            self.assertFalse(os.path.exists(expected_output_file))
```
Comment (collinss-jpl) [inline src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py:602]: This test is not adding much, and could be removed

**Review by collinss-jpl — CHANGES_REQUESTED** [review]
The approach looks solid, but something looked off about the last test in `_validate_output`.

Also, a general finding: everywhere you've used the term "burst" should be changed to "band", since that is what is indicated by the "B01_WTR", "B02_BWTR", etc...

File: src/opera/pge/dswx_s1/dswx_s1_pge.py (hunk 6 above)
Comment (JimHofman) [reply src/opera/pge/dswx_s1/dswx_s1_pge.py:171]: It's verifying that all the values are the same in the list.  Based on previous comments I imagine it should be called 'band_type' (i.e. B01_WTR, B02_BWTR or B03_CONF) or perhaps just 
'band'.

File: src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py (hunk 9 above)
Comment (JimHofman) [reply src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py:179]: In this case output_dir needs the complete path.  The relative path does not work on the 'os' commands.

File: src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py (hunk 10 above)
Comment (JimHofman) [reply src/opera/test/pge/dswx_s1/test_dswx_s1_pge.py:600]: I'll look at this again.  You're right the test is awkward.  I'll make it more clear what it's intended to do.  Thanks for the thorough review.

**Review by collinss-jpl — APPROVED** [review]

**Pull request merged**